source/slang/slang-options.cpp

```diff
@@ -1459,14 +1459,21 @@ struct OptionsParser
             // with an entry point, since the case of a single entry
             // point was handled above, and the user is expected to
             // follow the ordering rules when using multiple entry points.
-            //
-            // TODO(DG): check if output format is left unassociated
-            // If (known format for that output && that format supports whole program) {
-            //    set isWholeProgram bit on the TargetRequest }
-            // else { diagnose what happened }
             if( rawOutput.entryPointIndex == -1 )
             {
-                sink->diagnose(SourceLoc(), Diagnostics::cannotMatchOutputFileToEntryPoint, rawOutput.path);
+                if (rawOutput.targetIndex != -1 )
+                {
+                    auto outputFormat = rawTargets[rawOutput.targetIndex].format;
+                    switch (outputFormat)
+                    {
+                    case CodeGenTarget::CPPSource:
+                        rawOutput.isWholeProgram = true;
+                        break;
+                    default:
+                        sink->diagnose(SourceLoc(), Diagnostics::cannotMatchOutputFileToEntryPoint, rawOutput.path);
+                        break;
+                    }
+                }
```

Comment:
This new logic probably deserves some commenting. Most importantly, we should probably be noting that the `switch` is here to check whether the given target format supports multiple entry points in the same output file, and we should expand the list of supported cases here as/when additional targets add such support.